submdspan_future/submdspan-future.md

```diff
@@ -718,585 +755,803 @@ In [version.syn], increase the value of the `__cpp_lib_submdspan` macro by repla
+
+* [2.2]{.pnum} let `rank` be the number of elements in `slices`;
+
+* [2.3]{.pnum} let $s_k$ be the $k^{th}$ element of `slices`;
+  
+* [2.4]{.pnum} let $S_k$ be the type of $s_k$; and
+
+* [2.5]{.pnum} let  _`map-rank`_ be an `array<size_t, rank>` such that for each $k$ in the range $[0,$ `rank`$)$, _`map-rank`_`[`$k$`]` equals:
+
+    * [2.5.1]{.pnum} `dynamic_extent` if $S_k$ models `convertible_to<index_type>`,
+
+    * [2.5.2]{.pnum} otherwise, the number of types $S_j$ with $j < k$ that do not model `convertible_to<index_type>`.
+
+:::
+
+## Add section [mdspan.sub.slices], "Slices definitions and canonicalization"
+
+> Add a new section [mdspan.sub.slices], "Slices definitions and canonicalization," after [mdspan.sub.overview] and before [mdspan.sub.helpers].  The new section has the following contents.
 
 ::: add
-```
-template<class IndexType, class S>
-constexpr auto @_canonical-ice_@(S s);
-```
+[1]{.pnum} Given a signed or unsigned integer `IndexType`, a type $S$ is a *`submdspan` slice type for `IndexType`* if one of the following holds:
+
+* [1.1]{.pnum} `is_convertible_v<`$S$`, full_extent_t>` is `true`;
+
+* [1.2]{.pnum} `is_convertible_v<`$S$`, IndexType>` is `true`;
+
+* [1.3]{.pnum} $S$ is a specialization of `strided_slice` and `is_convertible_v<`$X$`, IndexType>` is true for $X$ denoting
+  $S$`::offset_type`, $S$`::extent_type` and $S$`::stride_type`; or
+
+* [1.4]{.pnum} all of the following hold:
+
+    * [1.4.1]{.pnum} the declaration `auto [L, U] = std::move(s);` is well-formed for some object `s` of type `S`,
+
+    * [1.4.2]{.pnum} `is_convertible_v<decltype(std::move(L))>, IndexType>` is `true`, and
+
+    * [1.4.3]{.pnum} `is_convertible_v<decltype(std::move(U))>, IndexType>` is `true`.
+
+[2]{.pnum} Given a signed or unsigned integer `IndexType`, a type $S$ is a *canonical `submdspan` index type for `IndexType`* if $S$ is either `IndexType` or `constant_wrapper<v, IndexType>` for some `v` of type `IndexType`, such that `v` is greater than or equal to zero.
+
+[3]{.pnum} Given a signed or unsigned integer `IndexType`, a type $S$ is a *canonical `submdspan` slice type for `IndexType`* if exactly one of the following is `true`:
+
+* [3.1]{.pnum} $S$ is `full_extent_t`;
+
+* [3.2]{.pnum} $S$ is a canonical `submdspan` index type for `IndexType`; or
+
+* [3.3]{.pnum} $S$ is a specialization of `strided_slice` where all of the following hold:
+
+    * [3.3.1]{.pnum} $S$`::offset_type`, $S$`::extent_type`, and $S$`::stride_type` are all canonical `submdspan` index types for `IndexType`; and
+
+    * [3.3.2]{.pnum} if $S$`::stride_type` and $S$`::extent_type` are both specializations of  `constant_wrapper`,
+      then $S$::`stride_type::value` is greater than zero.
+
+[4]{.pnum} Let `IndexType` be a signed or unsigned integer type.  A type `S` is a *collapsing slice type* if `S` is a canonical `submdspan` index type for `IndexType`.  An object `s` of type `S` is a *collapsing slice* if `S` is a collapsing slice type.

```

Comment:
I do not really like the introduction of `collapsing slice type` and `unit slice type`, we do not use them anywhere so they are just necessary.  We should apply Occam's razor here, and remove there.

I would much prefer saying:
```
Let `IndexType` be a signed or unsigned integer type , an object `s` is a *collapsing slice* if it's type ....
``

We could also free this definition from index type, if we say that it is *collapsing slice* if it's not specialization of strided_slice or full_extent.